synapse/federation/sender/__init__.py

```diff
@@ -560,3 +577,35 @@ async def get_replication_rows(
         # Dummy implementation for case where federation sender isn't offloaded
         # to a worker.
         return [], 0, False
+
+    async def _wake_destinations_needing_catchup(self):
+        """
+        Wakes up destinations that need catch-up and are not currently being
+        backed off from.
+        Does so in a slow way (one every 5 seconds) to reduce load spikes.
+        """
+
+        last_processed = None  # type: Optional[str]
+
+        while True:
+            destinations_to_wake = await self.store.get_catch_up_outstanding_destinations(
+                last_processed
+            )
+
+            if not destinations_to_wake:
+                # finished waking all destinations!
+                break
+
+            for destination in destinations_to_wake:
+                if self._federation_shard_config.should_handle(
+                    self._instance_name, destination
+                ):
```

Comment:
This logic seems to be duplicated from `wake_destination`. What was the goal here? Is it just to avoid the logging?

If there is multiple federation senders than they'll each run this code, but only handle the destinations they're interested in? It might be clearer to pre-filter the list, like:

https://github.com/matrix-org/synapse/blob/58f61f10f780a5f9e6be99f4072c24442594d597/synapse/federation/sender/__init__.py#L335-L340

Reply:
Yes, just to avoid the logging (because it'd be very confusing to have a federation sender claim it was waking a destination that it didn't) [and we also avoid the sleep].

Pre-filtering: no strong opinions there, fine by me — it's an extra pass and list allocation but only a small list so should be OK